src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -739,6 +740,93 @@ class StringGenerator : public Generator<std::string> {
     const size_t _alphabetLength;
 };
 
+/** `{^ChooseFromDataset:{...}` */
+class RandomStringFromDataset : public Generator<std::string> {
+
+public:
+    RandomStringFromDataset(const Node& node, GeneratorArgs generatorArgs)
+        : _rng{generatorArgs.rng},
+          _id{generatorArgs.actorId},
+          _path{node["path"].maybe<std::string>().value()} {
+        loadDataset();
+    }
+
+private:
+    void loadDataset() {
+        {
+            if (_path.empty()) {
+                BOOST_THROW_EXCEPTION(
+                    InvalidValueGeneratorSyntax("ChooseFromDataset requieres non-empty path"));
+            }
+
+            /* If path the doesn't exist in memory we load it.
+               To avoid reading and storing the datasets multiple times
+               we check if the path already exists in the map (as key). If
+               it doesn't exist, we open the file and store its contents.
+               If the number of items is more than 0, it means the data is
+               loaded, so we can skip everything, including the mutex.
+            */
+
+            if (_all_datasets.count(_path) > 0) {
+                return;
+            }
+
+            {
+                std::mutex _dataset_mutex;
```

Comment:
I might be missing something, but is this lock being declared locally on the stack? So every thread that construct a `RandomStringFromDataset` will be creating a separate mutex, locking its own lock, and getting no cross-thread coordination? I think the mutex should be a shared static var

Reply:
Oh, yes, that is a mistake when I was restructuring and moving variables. Will revert it to shared static.